mod-clans/src/main/java/gg/solarmc/loader/clans/Clan.java

```diff
@@ -0,0 +1,425 @@
+
+        receiver.setCachedClan(this);
+
+        return getClanMembers(transaction);
+    }
+
+    /**
+     * Removes a member from the clan.
+     * @param transaction The tx
+     * @param player player to remove from the clan
+     * @return Accurate set of ClanMembers post transaction
+     * @throws IllegalStateException if member already has a clan
+     * @throws IllegalArgumentException if member tried to remove was the leader of the clan, or if they are not
+     * present in the clan
+     */
+    public Set<ClanMember> removeClanMember(Transaction transaction, SolarPlayer player) {
+        ClanDataObject receiver = player.getData(ClansKey.INSTANCE);
+        if (receiver.getClan(transaction).isPresent()) throw new IllegalStateException("Receiver already has clan!");
+
+        ClansClanMembershipRecord rec = transaction.getProperty(DSLContext.class)
+                .fetchOne(CLANS_CLAN_MEMBERSHIP,CLANS_CLAN_MEMBERSHIP.CLAN_ID.eq(this.clanID).and(CLANS_CLAN_MEMBERSHIP.USER_ID.eq(receiver.getUserId())));
+
+        if (rec == null) throw new IllegalArgumentException("Member is not part of clan!");
+        if (receiver.isSimilar(leader)) throw new IllegalArgumentException("Tried to remove leader of clan!");
+
+        rec.delete();
+
+        receiver.setCachedClan(null);
+
+        return getClanMembers(transaction);
+    }
+
+    /**
+     * Adds clan as an ally
+     *
+     * @param transaction the tx
+     * @param receiver the receiver
+     * @return whether record was inserted or not.
+     * @throws IllegalArgumentException if receiver is the same gg.solarmc.loader.clans.Clan as this object
+     * @throws IllegalStateException if receiver or this object already has an ally
+     */
+    public boolean addClanAsAlly(Transaction transaction, Clan receiver) {
+        if (receiver.getID() == this.clanID) throw new IllegalArgumentException("Tried to mark clan ally as same clan!");
+        if (this.currentlyAlliedClanID().isPresent()) throw new IllegalStateException("Sender already have an ally!");
+        if (receiver.currentlyAlliedClanID().isPresent()) throw new IllegalStateException("Selected clan already has an ally!");
+
+        //note to aurium - these are now okay to do becaue we run the checks previous
+        int res = transaction.getProperty(DSLContext.class)
+                .insertInto(CLANS_CLAN_ALLIANCES)
+                .columns(CLANS_CLAN_ALLIANCES.CLAN_ID,CLANS_CLAN_ALLIANCES.ALLY_ID)
+                .values(this.clanID,receiver.getID())
+                .execute();
+
+        manager.insertAllianceCache(this.getID(),receiver.getID());
+
+        return res == 1;
+    }
+
+    /**
+     * Revokes ally. Can be called from either clan. Revokes alliance for both.
+     * @param transaction the tx
+     * @return whether insert was successful or not
+     * @throws IllegalStateException if the clan does not have an ally
+     */
+    public boolean revokeAlly(Transaction transaction) {
```

Comment:
This method should be somewhat re-design, both in terms of the specification and the implementation. Here's my idea:

```java
    /**
     * Revokes ally. Can be called from either clan. Revokes alliance for both.
     * @param transaction the tx
     * @return true if the allegiance was revoked, false if this clan did not have an ally
     */
    public boolean revokeAlly(Transaction transaction) {
      int updateCount = DELETE FROM clans_clan_alliances WHERE clan_id = this.clanID OR ally_id = this.clanID;
      return revoked;
    }
```

I'm mixing SQL and Java, but you get the idea. Run the DELETE, and check the update count to determine if the ally was revoked or if there was no ally.

You would also need to modify this example to call manager.invalidateAllianceCache if the ally was removed. For that, I recommend looking into SQL's RETURNING, so you can implement both the deletion and retrieval of the deleted clan ID in a single query:
https://mariadb.com/kb/en/delete/